harvester/harvester
# Support CPU Pinning

~~Blocked by https://github.com/harvester/harvester-installer/pull/756~~

**Problem:**
Implement CPU Pinning feature

**Solution:**
Introduce two endpoints:
- enable cpu manager: `/v1/harvester/nodes/{node_name}?action=enableCPUManager`, which turn cpu manager policy into `static` on the node. Then user could create vm with cpu pinning and deploy those vm to those nodes enable cpu manager.
- disable cpu manager: `/v1/harvester/nodes/{node_name}?action=disableCPUManager`, which turn cpu manager policy into `none`, note that if there is any vm with cpu pinning on the node that needed to disable cpu manager, an error raise `there should not be any running VMs with CPU pinning when disabling the CPU manager`.

for more details, please check the HEP: https://github.com/harvester/harvester/pull/5803

**Related Issue:**
related to https://github.com/harvester/harvester/issues/2305,

**Test plan:**
prerequisite: 
~~- build iso, note that you need to apply this patch
[0001-custom-installer.patch](https://github.com/user-attachments/files/16435446/0001-custom-installer.patch) after checkout to the branch which belongs to this PR. The patch change the harvester-installer version as I have to add some config files through harvester-installer while https://github.com/harvester/harvester-installer/pull/756 not merged yet.~~
- prepare 3 nodes harvester cluster
- create api key: click dashboard top right profile picture > Account & API Keys > Create API Key
- Use the following curl cmd to enable/disable cpu manager:
  a. enable cpu manager
  ```shell
      curl 'https://192.168.100.131/v1/harvester/nodes/{node_name}?action=enableCPUManager' \
      -X 'POST' \
      -H 'Authorization: Bearer <API_KEY>' \
      --insecure
  ```
  b. disable cpu manager
  ```shell
      curl 'https://192.168.100.131/v1/harvester/nodes/{node_name}?action=disableCPUManager' \
      -X 'POST' \
      -H 'Authorization: Bearer <API_KEY>' \
      --insecure
  ```

### Test enable/disable cpu pinning
1. enable cpu manager on harvester-node-0
2. Open k8s and there should have a job name start with `harvester-node-0-update-cpu-manager-` wait until the job done.
3. create vm named `test` with cpu pinning: go to dashboard and click `Virtual Machines` tab, click `Create` fill the Name, CPU, Memory, Volumes>Image field, and then click `Edit as YAML`. Add `dedicatedCpuPlacement: true` to `.spec.template.spec.domain.cpu`. And then click `Create`.
<img width="316" alt="image" src="https://github.com/user-attachments/assets/e35468ce-ff64-4849-a0ea-cebdf90c53c6">

4. make sure the cpu pinning works by using the following cmd
  ```
  kubectl exec {virt-launcher-pod-id} -- virsh dumpxml default_test | awk "/<cputune>/,/<\/cputune>/"
  ```
  the cmd should output something like belows, depends on the cpu size, there should have the corresponding `vcpupin` xml fields.
  ```
    <cputune>
      <vcpupin vcpu='0' cpuset='1'/>
      ...
    </cputune>
  ```
5. disable cpu manager in harvester-node-0, and verify user can't disable cpu manager when there is a cpu pinned vm in harvester-node-0.
An error raise: `admission webhook "validator.harvesterhci.io" denied the request: there should not be any running VMs with CPU pinning when disabling the CPU manager`
6. stop the vm `test`
7. disable cpu manager in harvester-node-0

###  Verify only one master node could enable/disable cpu manager
1. enable cpu manager on harvester-node-0, no need to wait until cpu manager job done, go to step 2.
2. enable cpu manager on harvester-node-1
In step 2, an error raised: `admission webhook "validator.harvesterhci.io" denied the request: there is another master node updating the CPU manager policy`. Enabling or disabling the CPU manager requires restarting the kubelet, which in turn necessitates restarting the rke2-server. Since we are using a three-node setup where all nodes have the master role, restarting all rke2-servers simultaneously could lead to the entire cluster going offline. Therefore, the CPU manager update process allows only one master node to update at a time.

### Verify there should be only one job enable/disable cpu manager per node
1. enable cpu manager on harvester-node-0, no need to wait until cpu manager job done, go to step 2.
2. enable or disable cpu manager on harvester-node-0
An error raise: `admission webhook "validator.harvesterhci.io" denied the request: there is other ongoing job updating cpu manager policy`

### Verify we can't enable cpu manager when it is enabled, and we can't disable cpu manager when it's disabled 
1. enable cpu manager on harvester-node-0, wait until cpu manager job done, go to step 2.
2. enable cpu manager on harvester-node-0, an error raises
`admission webhook "validator.harvesterhci.io" denied the request: same cpu manager policy`

### Test reboot
1. Create vm with cpu pinning
2. Reboot harvester node
3. After reboot, check rke2-server or rke2-agent  (depends on the node role) start up successfully and vm is running.

### Test upgrade path
Loop through all tests mentioned above under these 2 scenarios:
1. Test single node upgrade path: Prepare a one node harvester and upgrade to the version that include this feature
2. Test multi node upgrade path: Prepare multi nodes harvester cluster and upgrade to the version that include this feature.
```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Version
metadata:
  name: master
  namespace: harvester-system
spec:
  isoURL: http://192.168.100.1:8000/harvester-master-amd64.iso
---
apiVersion: harvesterhci.io/v1beta1
kind: Upgrade
metadata:
  annotations:
    harvesterhci.io/skip-version-check: "true"
    harvesterhci.io/skipWebhook: "true"
  generateName: hvst-upgrade-
  namespace: harvester-system
spec:
  version: "master"
  logEnabled: false
```
